### PR TITLE
Streaming data tests

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
Adds tests for composer + streaming for NLP datasets

TODO:
- [ ] figure out where these tests should actually read from (Q for Mihir)
- [x] unxfail all the scenarios as streaming bugs are fixed
- [x] mark as daily and remote
- [x] run locally on mac and gpu instance for world size 1/2 (ran `make test`, `make test-dist`, `WORLD_SIZE=2 make test-dist`, `make test-gpu`, `make test-dist-gpu`, and `WORLD_SIZE=2 make test-dist-gpu` on my Mac (excluding the gpu ones), and on a GPU instance)
- [x] In order to work on mac, this needs https://github.com/mosaicml/streaming/pull/97

# What issue(s) does this change relate to?
Closes [CO-1479](https://mosaicml.atlassian.net/browse/CO-1479)

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1479]: https://mosaicml.atlassian.net/browse/CO-1479?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/fixtures/autouse_fixtures.py [hunk 1]
```diff
@@ -13,6 +13,11 @@
 from composer.utils import dist, reproducibility
 
 
+@pytest.fixture(autouse=True)
+def disable_tokenizer_parallelism():
+    os.environ['TOKENIZERS_PARALLELISM'] = 'false'
```
Comment (mvpatel2000) [inline tests/fixtures/autouse_fixtures.py:18]: What does this do? can you maybe leave a comment?

Comment (dakinggg) [reply tests/fixtures/autouse_fixtures.py]: added a comment. basically it just silences a warning from huggingface

File: tests/datasets/test_streaming_datasets_train.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    dataloader = DataLoader(streaming_dataset, batch_size=16, num_workers=num_workers, collate_fn=collator)
+
+    trainer = Trainer(model=model, train_dataloader=dataloader, max_duration='2ba', device=device)
+
+    trainer.fit()
+
+    # Necessary for some reason, otherwise streaming does not clean up properly, and tests fail
+    trainer.close()
+    if trainer.state.train_dataloader and trainer.state.train_dataloader._iterator is not None:  # type: ignore [reportGeneralTypeIssues]
+        trainer.state.train_dataloader._iterator._shutdown_workers()  # type: ignore [reportGeneralTypeIssues]
```
Comment (mvpatel2000) [inline tests/datasets/test_streaming_datasets_train.py:83]: Should `engine.close()` call shutdown on the workers?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM, but want to hold off until after remote / daily migration to minimize moving pieces

File: tests/datasets/test_streaming_datasets_train.py (hunk 2 above)
Comment (dakinggg) [reply tests/datasets/test_streaming_datasets_train.py:83]: probably. can't imagine a situation where the engine is closed but you don't want the dataloader workers shut down.

Comment (mvpatel2000) [reply tests/datasets/test_streaming_datasets_train.py:83]: Can we add it in this PR?

Comment (dakinggg) [reply tests/datasets/test_streaming_datasets_train.py:83]: sure, will add

Comment (dakinggg) [reply tests/datasets/test_streaming_datasets_train.py:83]: https://github.com/mosaicml/composer/pull/1842/commits/4919d5a3978ee6a163e590c23a83a2083b2c8012

**Review by mvpatel2000 — APPROVED** [review]
would like to get review from @nik-mosaic as well

File: tests/datasets/test_streaming_datasets_train.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    streaming = pytest.importorskip('streaming')
+    transformers = pytest.importorskip('transformers')
+    name_to_cls = {
+        'c4': streaming.text.c4.StreamingC4,
+        'pile': streaming.text.pile.StreamingPile,
+        'enwiki': streaming.text.enwiki.StreamingEnWiki
+    }
+
+    full_seed = seed + (num_workers + 1) * 10 + (world_size + 1) * 100 + (1 if device == 'cpu' else 2) * 1000
+    # This seed setting is necessary to prevent a shared memory collision due to a streaming bug
```
Comment (nik-mosaic) [inline tests/datasets/test_streaming_datasets_train.py:53]: This exact seed???

**Review by nik-mosaic — APPROVED** [review]
Looks good, with the assumption that any metrics tests will be separate --- this is simply a matter of testing that the calls to train.fit() succeed.

File: tests/datasets/test_streaming_datasets_train.py (hunk 3 above)
Comment (nik-mosaic) [reply tests/datasets/test_streaming_datasets_train.py:53]: Resolving for now, but please follow up @dakinggg

Comment (dakinggg) [reply tests/datasets/test_streaming_datasets_train.py:53]: Not this exact seed. But the seeds needed to be unique across all parametrizations of this test. I think this has been fixed in the latest streaming, and I made a JIRA to clean up this test with the latest version of streaming
